### ISSUE TITLE
Startup time is now infuriatingly slow

### ISSUE THREAD

**Issue by rozyczko** [SasView/sasview#2228]
It now takes about 30 sec to start SasView from the dev build `python run.py` command.

Quick look shows that importing GenericScatteringCalculator and InversionPerspective takes a bit too long (16s and 11s respectively).
Let's have a look at why...

**Comment by rozyczko** [comment]
So it is `numba` initialization in these two modules which takes forever.
Removing numba from the env brings startup time down to ~5sec.

Is this what we are going to live with?

**Comment by butlerpd** [comment]
Even 5s seems like a long time I have to say. 30 is not "infuriatingly" or even "excruciating." "unacceptable" comes to mind 😄  Is there no way to load things in the background?

**Comment by smk78** [comment]
It's taking ~30s to start release 5.0.5 on my desktop here too. So it's not a dev build thing.

**Comment by smk78** [comment]
Are there numba.njit decorators in the offending perspectives? See https://stackoverflow.com/questions/73568079/python-numba-compiled-functions-import-very-slow Apparently you can cache these, though that leads to another potential issue...

**Comment by pkienzle** [comment]
One problem: numba is spitting out 17k lines of debug info to the sasview logs when started using run.py. This might have poor interactions with the slow filesystem on windows. You can suppress them using:
```diff
diff --git a/src/sas/logger_config.py b/src/sas/logger_config.py
index 7e93ae501..d9e4a5b38 100644
--- a/src/sas/logger_config.py
+++ b/src/sas/logger_config.py
@@ -27,6 +27,7 @@ class SetupLogger(object):
             logging.captureWarnings(True)
             logger = logging.getLogger(self.name)
         logging.getLogger('matplotlib').setLevel(logging.WARN)
+        logging.getLogger('numba').setLevel(logging.WARN)
         return logger
 
     def config_development(self):
@@ -37,6 +38,7 @@ class SetupLogger(object):
         self._update_all_logs_to_debug(logger)
         logging.captureWarnings(True)
         logging.getLogger('matplotlib').setLevel(logging.WARN)
+        logging.getLogger('numba').setLevel(logging.WARN)
         return logger
```

**Comment by pkienzle** [comment]
After debug suppress I can use run.py to time with and without caching (should work on mac, linux, and gitbash on windows):
```
$ time python run.py -c "import sas.sascalc.calculator.sas_gen"
...
5.55s user 0.42s system 110% cpu 5.386 total  # first time
1.58s user 0.37s system 144% cpu 1.347 total  # subsequent times
```

If I then modify anything in the file the time jumps back to 5.5s the first time then 1.5s on subsequent times.

The numba code is big. I was not able to figure out how they are generating the hash key for the cache by simple inspection. It could be something simple like file hash plus function signature. Quick to hash but error prone: if the generated code is referencing imported code that has changed then the simple signature will not change even though the effective source has changed. Or it could determine the code that is to be compiled and hash that prior to calling llvm. Slower (because it is already doing a lot of the work required for the jit) but less error prone.

There may also be problems with a stale cache and a new numba. They might already defend against that by including the numba version in the hash. Same for python version number in case the llvm code depends on the python ABI. And a numpy version number because numba has to interact with internal vector representations. Caching is hard…

Instead caching, you can put the just-in-time into jit and delay compiling until you actually need to run the code. A delay of a few seconds the first time you do a long calculation will not feel as bad as a long delay preparing code for all perspectives on startup.

**Comment by pkienzle** [comment]
Note: above tests run on "master" not on "main", which exists because … reasons.

The `python run.py -c ...` magic does not work with the qt gui.

**Comment by rozyczko** [comment]
> One problem: numba is spitting out 17k lines of debug info to the sasview logs when started using run.py. This might have poor interactions with the slow filesystem on windows. You can suppress them using:

I don't see those debug lines in the logs, at least not on Windows.
It would be better to address logging issues in general, as requested in #2230 rather than just add the quoted lines, I think.

**Comment by pbeaucage** [comment]
Was this fixed in #2275 or should we leave this issue open?
